File: 3620-MaximumNumberOfDistinctElementsAfterOperations/3620-MaximumNumberOfDistinctElementsAfterOperations.py
# Last updated: 1/12/2026, 3:13:04 PM
class Solution:
    def maxDistinctElements(self, nums: List[int], k: int) -> int:
        
        # sort the array
        nums.sort()

        count = 0
        curr = nums[0]-k-1

        for num in nums:

            # set it equal to minimum of the last thing we set + 1
            # and num - k.
            attempt = max(curr + 1, num - k)

            # if that is less than num + k, update count
            if attempt <= num + k:

                count += 1
                curr = attempt

        return count

File: 3620-MaximumNumberOfDistinctElementsAfterOperations/3620-MaximumNumberOfDistinctElementsAfterOperations.cpp
// Last updated: 1/12/2026, 3:13:08 PM
class Solution {
public:
    int maxDistinctElements(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
    
        int last_selection = INT_MIN;
        int ans = 0;
        for(int i = 0; i < n; i++) {
            int min = nums[i] - k;
            int max = nums[i] + k;

            if(min > last_selection) {
                last_selection = min;
                ans++;
            }
            else if(min <= last_selection && max > last_selection) {
                last_selection++;
                ans++;
            }
        }
        return ans;
    }
};